spring-cloud-gcp-storage/src/main/java/org/springframework/cloud/gcp/storage/GoogleStorageResource.java

```diff
@@ -0,0 +1,332 @@
+	 */
+	public Bucket createBucket() {
+		return this.storage.create(BucketInfo.newBuilder(this.bucketName).build());
+	}
+
+	/**
+	 * Returns the {@link Bucket} associated with the resource.
+	 * @return the bucket if it exists, or null otherwise
+	 */
+	public Bucket getBucket() {
+		return this.storage.get(this.bucketName);
+	}
+
+	/**
+	 * Checks for the existence of the {@link Bucket} associated with the resource.
+	 * @return true if the bucket exists
+	 */
+	public boolean bucketExists() {
+		return getBucket() != null;
+	}
+
+	private Blob throwExceptionForNullBlob(Blob blob) throws IOException {
+		if (blob == null) {
+			throw new FileNotFoundException("The blob was not found: " + this.location);
+		}
+		return blob;
+	}
+
+	@Override
+	public File getFile() throws IOException {
+		throw new UnsupportedOperationException(
+				getDescription() + " cannot be resolved to absolute file path");
+	}
+
+	@Override
+	public long contentLength() throws IOException {
+		return throwExceptionForNullBlob(getBlob()).getSize();
+	}
+
+	@Override
+	public long lastModified() throws IOException {
+		return throwExceptionForNullBlob(getBlob()).getUpdateTime();
+	}
+
+	/**
+	 * Creates a file in the same bucket as this one. It inherits {@code autoCreateFiles}
+	 * from this object.
+	 *
+	 * <p>The file is created if it didn't exist in Google Cloud Storage.
+	 * @param relativePath the URL to a Google Cloud Storage file
+	 * @return the {@link GoogleStorageResource} object for the created file
+	 * @throws IOException
+	 */
+	@Override
+	public GoogleStorageResource createRelative(String relativePath) throws IOException {
+		GoogleStorageResource blobResource = new GoogleStorageResource(
+				this.storage, this.location.resolve(relativePath).toString(), this.autoCreateFiles);
+
+		if (!blobResource.exists()) {
+			blobResource.createBlob();
+		}
+
+		return blobResource;
+	}
+
+	@Override
+	public String getFilename() {
+		return isBucket() ? this.bucketName : this.blobName;
+	}
+
+	@Override
+	public String getDescription() {
+		return this.location.toString();
+	}
+
+	@Override
+	public InputStream getInputStream() throws IOException {
+		return Channels.newInputStream(throwExceptionForNullBlob(getBlob()).reader());
+	}
+
+	@Override
+	public boolean isWritable() {
+		return !isBucket() && (this.autoCreateFiles || exists());
+	}
+
+	/**
+	 * Returns the output stream for a Google Cloud Storage file.
+	 * @return the object's output stream or {@code null} if the object doesn't exist and cannot be
+	 * created
+	 * @throws IOException
+	 */
+	@Override
+	public OutputStream getOutputStream() throws IOException {
+		Blob blob = getBlob();
+
+		if (blob == null || !blob.exists()) {
+			if (!this.autoCreateFiles) {
+				throw new FileNotFoundException("The blob was not found: " + this.location);
+			}
+
+			blob = createBlob();
+		}
+
+		return Channels.newOutputStream(blob.writer());
+	}
+
+	public String getBlobName() {
```

Comment:
Why not just using `getFilename()`?

Reply:
Implementation of the `getFilename()` might return the bucket name, if the resource represents a bucket. So, they're not exactly the same.

Reply:
So again, if this resource represents a bucket, this would return null?
Another pointer why this probably shouldn't represent both a bucket and blob at the same time...